cmd/clusterctl/client/cluster/installer.go

```diff
@@ -90,6 +111,63 @@ func installComponentsAndUpdateInventory(components repository.Components, provi
 	return providerInventory.Create(inventoryObject)
 }
 
+// waitForReady waits till the installed components are ready.
+func (i *providerInstaller) waitForReady(opts InstallOptions) error {
+	// If we dont have to wait for providers to be installed
+	// return early.
+	if !opts.WaitProviders {
+		return nil
+	}
+
+	log := logf.Log
+	log.Info("Waiting for providers to be available...")
+
+	return i.waitManagerDeploymentsReady(opts)
+}
+
+// waitDeploymentsReady waits till the installed deployments are ready.
+func (i *providerInstaller) waitManagerDeploymentsReady(opts InstallOptions) error {
+	for _, components := range i.installQueue {
+		for _, obj := range components.Objs() {
+			if util.IsManagerDeployment(obj) {
+				if err := i.waitDeploymentReady(obj, opts); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (i *providerInstaller) waitDeploymentReady(deployment unstructured.Unstructured, opts InstallOptions) error {
+	log := logf.Log
+	err := wait.Poll(100*time.Millisecond, opts.WaitProvidersTimeout, func() (bool, error) {
+		c, err := i.proxy.NewClient()
+		if err != nil {
+			return false, err
+		}
+		key := client.ObjectKey{
+			Namespace: deployment.GetNamespace(),
+			Name:      deployment.GetName(),
+		}
+		dep := &appsv1.Deployment{}
+		if err := c.Get(context.TODO(), key, dep); err != nil {
+			return false, err
+		}
+		for _, c := range dep.Status.Conditions {
+			if c.Type == appsv1.DeploymentAvailable && c.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err == wait.ErrWaitTimeout {
```

Comment:
I wonder if we should instead do something like:
```go
if err != nil {
    return errors.Wrap(err, fmt.Sprintf("failed to wait until Deployment %s/%s is ready", deployment.GetNamespace(), deployment.GetName()))
}
return nil
```

Advantages:
* We get a nice user-facing error message (independent of what exactly fails)
* I assume (but I'm not 100% sure) when we add a log here and return an error an upper layer will (hopefully) also log the error to the user and then we would have the error twice.

Reply:
Here is why i chose to not error out if waiting for deployments timed out:  
The primary goal of `clusterctl init` is to install the providers which, at this point, it finished successfully. I wasn't convinced that throwing an error here is the correct UX as we just timeout waiting for the providers to be ready. Technically, installation was a success. Hence, I was just looking to just throw a warning message to the user that we timed out waiting for the deployments.

That being said, since the wait feature is optional and behind flag it might be fine to throw an error if we timed out waiting for the providers. 

Which of these is better UX? 

Thoughts?

@fabriziopandini @vincepri

Reply:
I misread your code, so my arguments don't exactly make sense :)

Imho the `clusterctl init` is not successful when the deployments are not getting ready. But let's see what others think.

Reply:
Got it. I will refactor to error out. 🙂